### PR TITLE
Add support for CaDiCaL as eager BV SAT solver.

### PR DESCRIPTION
**Pull request by mpreiner**

### CODE REVIEW

File: src/prop/cadical.h [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    ~Statistics();
+  };
+
+  Statistics d_statistics;
+};
+
+}  // namespace prop
+}  // namespace CVC4
+
+#endif  // CVC4_USE_CRYPTOMINISAT
```
Comment (timothy-king) [inline src/prop/cadical.h]: CVC4_USE_CADICAL

File: src/prop/cadical.h [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  SatValue value(SatLiteral l) override;
+
+  SatValue modelValue(SatLiteral l) override;
+
+  unsigned getAssertionLevel() const override;
+
+  bool ok() const override;
+
+ private:
+  CaDiCaL::Solver* d_solver;
```
Comment (timothy-king) [inline src/prop/cadical.h]: unique_ptr ?

Comment (mpreiner) [reply src/prop/cadical.h]: Done.

File: src/prop/cadical.h (hunk 1 above)
Comment (mpreiner) [reply src/prop/cadical.h]: Done.

File: src/prop/cadical.h (hunk 2 above)
Comment (mpreiner) [reply src/prop/cadical.h]: Also added unique_ptr to the cryptominisat version (in this commit).

File: COPYING [hunk 3]
```diff
@@ -82,7 +82,8 @@ GMP) may (and do) remain under the more permissive modified BSD license.
 CVC4 optionally links against the following libraries:
 
   ABC                (https://bitbucket.org/alanmi/abc)
-  Cryptominisat4.0   (https://github.com/msoos/cryptominisat)
+  CaDiCaL            (https://github.com/arminbiere/cadical)
+  Cryptominisat      (https://github.com/msoos/cryptominisat)
```
Comment (4tXJ7f) [inline COPYING]: Nit: It is CryptoMiniSat

File: config/cadical.m4 [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  CADICAL_LDFLAGS="-L$CADICAL_HOME/build"
+
+else
+  AC_MSG_RESULT([no, user didn't request CaDiCaL])
+  with_cadical=no
+fi
+
+])# CVC4_CHECK_FOR_CADICAL
+
+# CVC4_TRY_STATIC_CADICAL_WITH(LIBS)
```
Comment (4tXJ7f) [inline config/cadical.m4]: Nit: The function is called `CVC4_TRY_CADICAL`.

File: config/cadical.m4 [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+      [[CaDiCaL::Solver test();]])], [CADICAL_LIBS="-lcadical"],
+    [CADICAL_LIBS=])
+
+  LDFLAGS="$cvc4_save_LDFLAGS"
+  CPPFLAGS="$cvc4_save_CPPFLAGS"
+  LIBS="$cvc4_save_LIBS"
+
+  AC_LANG_POP([C++])
+fi
+])# CVC4_TRY_CADICAL_WITH
```
Comment (4tXJ7f) [inline config/cadical.m4]: Nit: The function is called `CVC4_TRY_CADICAL`.

File: contrib/get-cadical [hunk 6]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+#
+source "$(dirname "$0")/get-script-header.sh"
+
+if [ -e cadical ]; then
+  echo 'error: file or directory "cadical" exists; please move it out of the way.' >&2
+  exit 1
+fi
+
+git clone https://github.com/arminbiere/cadical cadical
```
Comment (4tXJ7f) [inline contrib/get-cadical]: For reproducability of results/bugs, we should maybe get a specific commit? At least, we should be able to print the commit used.

File: config/cadical.m4 [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    AC_MSG_RESULT([yes, $CADICAL_HOME])
+  else
+    AC_MSG_RESULT([no])
+  fi
+
+  if ! test -d "$CADICAL_HOME" || ! test -e "$CADICAL_HOME/build/libcadical.a" ; then
+    AC_MSG_FAILURE([either $CADICAL_HOME is not a CaDiCaL source tree or it's not yet built $CADICAL_HOME/build/libcadical.a])
+  fi
+
+  CPPFLAGS="$CPPFLAGS -I$CADICAL_HOME/src"
```
Comment (4tXJ7f) [inline config/cadical.m4]: It looks like this should be together with lines 76/77. You are adding the folder as an include path in configure.ac, so no need to modify the `CPPFLAGS` globally.

File: src/options/options_handler.cpp [hunk 8]
```diff
@@ -1042,32 +1042,38 @@ void OptionsHandler::abcEnabledBuild(std::string option, std::string value)
...
 {
-#ifndef CVC4_USE_CRYPTOMINISAT
-  if(value) {
+#if !defined(CVC4_USE_CRYPTOMINISAT) && !defined(CVC4_USE_CADICAL)
+  if (value)
+  {
     std::stringstream ss;
-    ss << "option `" << option << "' requires an cryptominisat-enabled build of CVC4; this binary was not built with cryptominisat support";
+    ss << "option `" << option
+       << "' requires a CVC4 to be built with Cryptominisat or CaDiCaL";
```
Comment (4tXJ7f) [inline src/options/options_handler.cpp]: CryptoMiniSat :)

File: src/options/options_handler.cpp [hunk 9]
```diff
@@ -1042,32 +1042,38 @@ void OptionsHandler::abcEnabledBuild(std::string option, std::string value)
...
 {
-#ifndef CVC4_USE_CRYPTOMINISAT
-  if(!value.empty()) {
+#if !defined(CVC4_USE_CRYPTOMINISAT) && !defined(CVC4_USE_CADICAL)
+  if (!value.empty())
+  {
     std::stringstream ss;
-    ss << "option `" << option << "' requires an cryptominisat-enabled build of CVC4; this binary was not built with cryptominisat support";
+    ss << "option `" << option
+       << "' requires a CVC4 to be built with Cryptominisat or CaDiCaL";
```
Comment (4tXJ7f) [inline src/options/options_handler.cpp]: CryptoMiniSat :)

File: src/options/options_handler.cpp [hunk 10]
```diff
@@ -1086,7 +1092,8 @@ theory::bv::SatSolverMode OptionsHandler::stringToSatSolver(std::string option,
 
     if (options::bitblastMode() == theory::bv::BITBLAST_MODE_LAZY &&
         options::bitblastMode.wasSetByUser()) {
-      throw OptionException(std::string("Cryptominsat does not support lazy bit-blsating. \n\
+      throw OptionException(
+          std::string("Cryptominsat does not support lazy bit-blasting. \n\
```
Comment (4tXJ7f) [inline src/options/options_handler.cpp]: CryptoMiniSat or at least Cryptominisat (instead of Cryptominsat)

File: src/prop/cadical.cpp [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+  d_solver->add(0);
+  d_solver->add(-toCadicalVar(d_false));
+  d_solver->add(0);
+}
+
+CadicalSolver::~CadicalSolver() {}
+
+ClauseId CadicalSolver::addClause(SatClause& clause, bool removable)
+{
+  for (SatLiteral lit : clause)
```
Comment (4tXJ7f) [inline src/prop/cadical.cpp]: `const SatLiteral&`? Shouldn't make a huge difference but nevertheless.

File: src/prop/cadical.cpp [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+  d_solver->set("quiet", 1);  // CaDiCaL is verbose by default
+  d_solver->add(toCadicalVar(d_true));
+  d_solver->add(0);
+  d_solver->add(-toCadicalVar(d_false));
+  d_solver->add(0);
+}
+
+CadicalSolver::~CadicalSolver() {}
+
+ClauseId CadicalSolver::addClause(SatClause& clause, bool removable)
```
Comment (4tXJ7f) [inline src/prop/cadical.cpp]: Why isn't clause `const SatClause&`? (I know, this method is inherited from SatSolver but we should probably change this at some point)

File: src/prop/cadical.cpp [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+  for (SatLiteral lit : clause)
+  {
+    d_solver->add(toCadicalLit(lit));
+  }
+  d_solver->add(0);
+  ++d_statistics.d_numClauses;
+  return ClauseIdError;
+}
+
+ClauseId CadicalSolver::addXorClause(SatClause& clause,
```
Comment (4tXJ7f) [inline src/prop/cadical.cpp]: Same as above: `const SatClause&`?

File: config/cadical.m4 (hunk 7 above)
Comment (mpreiner) [reply config/cadical.m4]: No,  76/77 is only set temporarily for the `AC_LINK_IFELSE` check. After the check the flags get restored.

File: COPYING (hunk 3 above)
Comment (mpreiner) [reply COPYING]: Done.

File: config/cadical.m4 (hunk 4 above)
Comment (mpreiner) [reply config/cadical.m4]: Done.

File: config/cadical.m4 (hunk 5 above)
Comment (mpreiner) [reply config/cadical.m4]: Done.

File: contrib/get-cadical (hunk 6 above)
Comment (mpreiner) [reply contrib/get-cadical]: Done, added specific commit.

File: src/options/options_handler.cpp (hunk 8 above)
Comment (mpreiner) [reply src/options/options_handler.cpp]: Done.

File: src/options/options_handler.cpp (hunk 9 above)
Comment (mpreiner) [reply src/options/options_handler.cpp]: Done.

File: src/options/options_handler.cpp (hunk 10 above)
Comment (mpreiner) [reply src/options/options_handler.cpp]: Done.

File: src/prop/cadical.cpp (hunk 12 above)
Comment (mpreiner) [reply src/prop/cadical.cpp]: Probably for no specific reason. We can clean this up in a separate PR later.

File: src/prop/cadical.cpp (hunk 11 above)
Comment (mpreiner) [reply src/prop/cadical.cpp]: Done.

File: src/prop/cadical.cpp (hunk 13 above)
Comment (mpreiner) [reply src/prop/cadical.cpp]: Ditto.

**Pull request merged**